pkg/kafka/policy.go

```diff
@@ -281,9 +222,42 @@ func (req *RequestMessage) ruleMatches(rule api.PortRuleKafka) bool {
 // rules. The function will return true if the policy allows the message,
 // otherwise false is returned.
 func (req *RequestMessage) MatchesRule(rules []api.PortRuleKafka) bool {
+	reqTopicsMap := map[string]bool{}
+
+	// Maintain a map of all topics in the request
+	// Only if all the topics in a request are
+	// allowed by the rules, we should allow a request.
+	topics := req.GetTopics()
+	for _, topic := range topics {
+		reqTopicsMap[topic] = true
+	}
+
 	for _, rule := range rules {
-		if req.ruleMatches(rule) {
-			return true
+		switch len(topics) {
+		case 0:
+			if req.ruleMatches(rule, "") {
+				return true
+			}
+		default:
+			if rule.Topic != "" {
+				_, ok := reqTopicsMap[rule.Topic]
+				if ok && req.ruleMatches(rule, rule.Topic) {
+					delete(reqTopicsMap, rule.Topic)
+					// If all topics in a request are allowed by the
+					// set of rules we have traversed so far, along with the other
+					// fields in the request, the request
+					// should be allowed.
+					if len(reqTopicsMap) == 0 {
+						return true
+					}
+				}
+			} else {
+				// topic in rule is empty,
+				// check if other fields of the rule match.
+				if req.ruleMatches(rule, "") {
```

Comment:
This is definitely wrong as it returns `true` even if some topics may have been left unmatched.

I would make `ruleMatches()` aware of the `reqTopicsMap` instead and delete found topics inside of the function. That way you can radically simplify the logic outside and make it identical for the cases where you are looking for topics or not.

Reply:
This code is hit when
```if rule.Topic == "" {```
So this means we allow all topics by policy?

I can simplify if you like, but I don't think the logic is wrong?

Reply:
How about this:
```
//simplified
	count := 0
	for _, topic := range topics {
		for _, rule := range rules {
			switch len(topics) {
			case 0:
				if req.ruleMatches(rule, "") {
					return true
				}
			default:
				if req.ruleMatches(rule, topic) {
					count++
					break
				}
			}
		}
	}
	
	if (count == len(topics)) {
		return true
	}

	return false
```

I was trying to optimize the outer` topic` loop by caching it. But if this looks more simplified, then I can do this instead.

Reply:
The above doesn't work as it could count the same topic twice.

Before I was worried about a single rule with a topic wildcard to only apply to some of the topics in the message but I just realized that topic is the only field that can occur multiple times for now. So it is fine as-is.